src/integer/z/sample/binomial.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright © 2023 Niklas Siemer
+//
+// This file is part of qFALL-math.
+//
+// qFALL-math is free software: you can redistribute it and/or modify it under
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! This module contains algorithms for sampling
+//! according to the binomial distribution.
+
+use crate::{error::MathError, integer::Z, rational::Q, utils::sample::binomial::sample_binomial};
+
+impl Z {
+    /// Chooses a [`Z`] instance according to the binomial distribution
+    /// parameterized by `n` and `p`.
+    ///
+    /// Parameters:
+    /// - `n`: specifies the number of trials
+    /// - `p`: specifies the probability of success
+    ///
+    /// Returns a fresh [`Z`] instance with a value sampled
+    /// according to the binomial distribution or a [`MathError`]
+    /// if `n < 1`, `p <= 0`, `p >= 1`, or `n` does not fit into an [`i64`].
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer::Z;
+    ///
+    /// let sample = Z::sample_binomial(2, 0.5).unwrap();
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Returns a [`MathError`] of type [`InvalidIntegerInput`](MathError::InvalidIntegerInput)
+    /// if `n < 1`.
+    /// - Returns a [`MathError`] of type [`InvalidInterval`](MathError::InvalidInterval)
+    /// if `p <= 0` or `p >= 1`.
+    /// - Returns a [`MathError`] of type [`ConversionError`](MathError::ConversionError)
+    /// if `n` does not fit into an [`i64`].
+    pub fn sample_binomial(n: impl Into<Z>, p: impl Into<Q>) -> Result<Self, MathError> {
+        let n: Z = n.into();
+        let p: Q = p.into();
+
+        let sample = sample_binomial(&n, &p)?;
+        Ok(Z::from(sample))
```

Comment:
Please consider the following change, just now came to my mind: I would let the binomial sampler (the intern function return a `Z` value rather than a `u64` -> This makes changing it later much easier as only the underlying function has to change and not each function calling it

Reply:
That's a change, I'm going to make once we want to sample values larger than u64::MAX. Currently, it is just more efficient to return the u64 value and transform it into the needed value.